pyogrio/_io.pyx

```diff
@@ -1914,22 +1948,27 @@ def ogr_write(
         gdal_tz_offsets = {}
 
     ### Setup up dataset and layer
-    if not encoding:
-        encoding = locale.getpreferredencoding()
-
     layer_created = create_ogr_dataset_layer(
         path, layer, driver, crs, geometry_type, encoding,
         dataset_kwargs, layer_kwargs, append,
         dataset_metadata, layer_metadata,
         &ogr_dataset, &ogr_layer,
     )
 
+    # Now the dataset and layer have been created, we can properly determine the
+    # encoding. It is derived from the user, from the dataset capabilities / type,
+    # or from the system locale
+    encoding = (
+        encoding
+        or detect_encoding(ogr_dataset, ogr_layer)
+        or locale.getpreferredencoding()
```

Comment:
It looks like we always fall back to `locale.getpreferredencoding()` everywhere we are trying to get the encoding; what do you think about centralizing this in `detect_encoding`?

```Python
cdef detect_encoding(OGRDataSourceH ogr_dataset, OGRLayerH ogr_layer):
    ...

    if OGR_L_TestCapability(ogr_layer, OLCStringsAsUTF8):
        return 'UTF-8'

    driver = get_driver(ogr_dataset)
    encoding = get_encoding_for_driver(driver)

   return encoding or locale.getpreferredencoding()
```

Reply:
OK!